core/src/main/scala/kafka/log/LogLoader.scala

```diff
@@ -0,0 +1,526 @@
+import org.apache.kafka.common.errors.InvalidOffsetException
+import org.apache.kafka.common.utils.Time
+
+import scala.collection.{Seq, Set, mutable}
+
+case class LoadedLogOffsets(logStartOffset: Long,
+                            recoveryPoint: Long,
+                            nextOffsetMetadata: LogOffsetMetadata)
+
+/**
+ * @param dir The directory from which log segments need to be loaded
+ * @param topicPartition The topic partition associated with the log being loaded
+ * @param config The configuration settings for the log being loaded
+ * @param scheduler The thread pool scheduler used for background actions
+ * @param time The time instance used for checking the clock
+ * @param logDirFailureChannel The LogDirFailureChannel instance to asynchronously handle log
+ *                             directory failure
+ * @param hadCleanShutdown Boolean flag to indicate whether the associated log previously had a
+ *                         clean shutdown
+ * @param segments The LogSegments instance into which segments recovered from disk will be
+ *                 populated
+ * @param logStartOffsetCheckpoint The checkpoint of the log start offset
+ * @param recoveryPointCheckpoint The checkpoint of the offset at which to begin the recovery
+ * @param maxProducerIdExpirationMs The maximum amount of time to wait before a producer id is
+ *                                  considered expired
+ * @param leaderEpochCache An optional LeaderEpochFileCache instance to be updated during recovery
+ * @param producerStateManager The ProducerStateManager instance to be updated during recovery
+ */
+case class LoadLogParams(dir: File,
+                         topicPartition: TopicPartition,
+                         config: LogConfig,
+                         scheduler: Scheduler,
+                         time: Time,
+                         logDirFailureChannel: LogDirFailureChannel,
+                         hadCleanShutdown: Boolean,
+                         segments: LogSegments,
+                         logStartOffsetCheckpoint: Long,
+                         recoveryPointCheckpoint: Long,
+                         maxProducerIdExpirationMs: Int,
+                         leaderEpochCache: Option[LeaderEpochFileCache],
+                         producerStateManager: ProducerStateManager) {
+  val logIdentifier: String = s"[LogLoader partition=$topicPartition, dir=${dir.getParent}]"
+}
+
+/**
+ * This object is responsible for all activities related with recovery of log segments from disk.
+ */
+object LogLoader extends Logging {
+  /**
+   * Load the log segments from the log files on disk, and return the components of the loaded log.
+   * Additionally, it also suitably updates the provided LeaderEpochFileCache and ProducerStateManager
+   * to reflect the contents of the loaded log.
+   *
+   * This method does not need to convert IOException to KafkaStorageException because it is only
+   * called before all logs are loaded.
```

Comment:
Thise comment was trying to convey that in the context of the calling thread, this function doesn't convert the IOException to KafkaStorageException. In the recovery path, we always asynchronously delete segments. The IOException conversion happens only within the background work that's scheduled for segment deletions. https://github.com/apache/kafka/blob/193a9dfc44fa2c0df163a8518a804c4579d37e36/core/src/main/scala/kafka/log/Log.scala#L2374-L2389

In case this is misleading, does the below improved comment feel better to you?
```
In the context of the calling thread, this function does not need to convert
IOException to KafkaStorageException because it is only called before all logs are loaded.
```